presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -335,81 +342,123 @@ protected Object visitIsNotNullPredicate(IsNotNullPredicate node, Object context
         @Override
         protected Object visitSearchedCaseExpression(SearchedCaseExpression node, Object context)
         {
-            Expression resultClause = node.getDefaultValue().orElse(null);
+            Object defaultResult = processWithExceptionHandling(node.getDefaultValue().orElse(null), context);
+
+            List<WhenClause> whenClauses = new ArrayList<>();
             for (WhenClause whenClause : node.getWhenClauses()) {
-                Object value = process(whenClause.getOperand(), context);
-                if (value instanceof Expression) {
-                    // TODO: optimize this case
-                    return node;
+                Object whenOperand = processWithExceptionHandling(whenClause.getOperand(), context);
+                Object result = processWithExceptionHandling(whenClause.getResult(), context);
+
+                if (whenOperand instanceof Expression) {
+                    // cannot fully evaluate, add updated whenClause
+                    whenClauses.add(new WhenClause(
+                            toExpression(whenOperand, type(whenClause.getOperand())),
+                            toExpression(result, type(whenClause.getResult()))));
                 }
-
-                if (Boolean.TRUE.equals(value)) {
-                    resultClause = whenClause.getResult();
+                else if (Boolean.TRUE.equals(whenOperand)) {
+                    // condition is true, use this as defaultResult
+                    defaultResult = result;
                     break;
                 }
             }
 
-            if (resultClause == null) {
+            if (whenClauses.isEmpty()) {
+                return defaultResult;
+            }
+
+            Expression resultExpression = (defaultResult == null) ? null : toExpression(defaultResult, type(node));
```

Comment:
You could just do `resultExpression = toExpression(defaultResult, type(node))`. Having the optimized SearchedCaseExpression with a default case that might explicitly return null is not a big deal.